src/allmydata/test/web/test_web.py

```diff
@@ -759,12 +756,8 @@ def test_unknown_format(self):
         returned.
         """
         resource = self.resource()
-        self.assertIn(
-            "<title>Exception</title>",
+        with self.assertRaises(WebError):
             self.render(resource, t=["foo"]),
-        )
-        self.flushLoggedErrors(WebError)
-
```

Comment:
Hm.  Before, rendering the resource would respond with an error page.  Maybe that wasn't great behavior.  However, what happens now?  An exception is synchronously raised out of the render method and the request doesn't receive _any_ response?  That sounds worse.

Reply:
It will result in a 400 Bad Request:

```
$ http :3456?t=foo
HTTP/1.1 400 Bad Request
Content-Length: 26
Content-Type: text/plain;charset=utf-8
Date: Thu, 27 Aug 2020 15:41:10 GMT
Referrer-Policy: no-referrer
Server: TwistedWeb/20.3.0
X-Frame-Options: DENY

Unknown t value: 'foo'
```

Shouldn't that be good enough?

Reply:
Maybe so.  Is it tested anywhere?

Reply:
Updated the test case this way:

```python
    def test_unknown_format(self):
        """
        If a format is selected for which there is no renderer, an error is
        returned.
        """
        resource = self.resource()
        with self.assertRaises(WebError) as ex:
            self.render(resource, t=["foo"]),
        self.assertEquals(ex.exception.text, "Unknown t value: 'foo'")
        self.assertEquals(ex.exception.code, 400)
```

Reply:
As I understand it, this still doesn't make any assertions about what response is generated.  It just asserts that a `WebError` gets raised.

If "raise a WebError to render an error response" is the API then I guess this is fine.  Is that behavior tested somewhere else?

Reply:
GitHub is weird.  I posted a note in this thread from the PR "diff" view, which somehow caused a review to start.  Came back to the PR, saw that I had started a review on a PR that I myself had submitted, cancelled the "review", which made the review and the note I originally meant to post vanish.

Will try again later.

Reply:
My reading is that "raise a `WebError` when the requested format is not known to the `MultiFormatResource` instance in question".  I have not understood what turns a `WebError` into an HTTP 400 response, although that is what I have observed.

This is what I am trying to test:

https://github.com/tahoe-lafs/tahoe-lafs/blob/e17e9fa7a99af83657dc6f280d6698721b22cd04/src/allmydata/web/common_py3.py#L93-L99

In this case, shouldn't it be sufficient to check for the response code and error text that `WebError` carries?  Is there a better way to test this?

I have not found another place where this behavior is tested, at least directly.

Reply:
I think "WebError" here is really `twisted.web.error.Error` so yes it does appear to be "the API" (to raise things in `render` .. or at least some internal Twisted stuff does this too).

Reply:
I wasn't able to find any Twisted documentation for this feature.  The explicit "error handling" documentation for Twisted Web is https://twistedmatrix.com/documents/current/web/howto/web-in-60/error-handling.html (which, n.b., I wrote) and doesn't talk about this feature and the API documentation for `render` (https://twistedmatrix.com/documents/current/api/twisted.web.resource.IResource.html) only documents `UnsupportedMethod`.

Maybe we can submit a documentation PR upstream to Twisted for this?  Or at least file a ticket and link to it from the implementation.  I expect this is *probably* all fine and intended to be a public feature but it would be nice to have a few words to that effect somewhere.

Reply:
Actually, it looks like `WebError` is not `twisted.web.error.Error` but a custom Tahoe-LAFS exception, `allmydata.web.common_py3.WebError` which is a direct `Exception` subclass.

The behavior we get from Twisted is the catch-all "application code raises an exception I didn't recognize so I'm splatting out a pile of html".  This _also_ writes a traceback to the Twisted logging system.  Considering this case only means "the user made a bad request", having a traceback in the logs is probably just noise.

A preferable behavior would be to return an `twisted.web.resource.ErrorPage` that renders the error we want.  This will result in just the normal amount of http logging.

However, that doesn't really have much to do with the Nevow/Twisted Web porting going on here so I think it's out of scope for this PR.  What still seems like it should be in-scope is preserving the test coverage for the "exception" response being generated.  master@HEAD currently has this assertion and the PR changes the test to replace it with the weaker "`WebError` got raised" assertion.

Reply:
Some more background...

The way `WebError` works when Nevow is the app server is that Tahoe-LAFS registers a custom Nevow `ICanHandleException` provider.  This object gets to take a first look at any exceptions raised during rendering.  There is no equivalent Twisted Web feature.  This probably means that *all* Tahoe-LAFS `raise WebError(...)` needs to be given some treatment before we can switch the app server from Nevow to Twisted Web.

Reply:
Or put another way, porting `MultiFormatResourceTests.render` from Nevow to Twisted Web revealed a real incompatibility (while also removing our test coverage for rendering our resources using Nevow) and the test should not be changed to paper over this incompatibility - the behavior needs to be fixed so the test can continue to pass as previously written.

Reply:
I pushed a small change to switch MultiFormatResource from WebError to ErrorPage.

Reply:
Looks like pyrsistent packaging issue is yet to be fixed.  Will merge master which puts an upper bound on pyrsistent version.